katdal/chunkstore.py

```diff
@@ -19,25 +19,36 @@
 from __future__ import division
 
 import contextlib
+import functools
 
 import numpy as np
 import dask
 import dask.array as da
 
 
-class StoreUnavailable(OSError):
+class ChunkStoreError(Exception):
+    """"Base class for all standard ChunkStore errors."""
+
+
+class StoreUnavailable(OSError, ChunkStoreError):
     """Could not access underlying storage medium (offline, auth failed, etc)."""
 
 
-class ChunkNotFound(KeyError):
+class ChunkNotFound(KeyError, ChunkStoreError):
     """The store was accessible but a chunk with the given name was not found."""
 
 
-class BadChunk(ValueError):
+class BadChunk(ValueError, ChunkStoreError):
     """The chunk is malformed, e.g. bad dtype or slices, wrong buffer size."""
 
 
-def generate_chunks(shape, dtype, max_chunk_size, dims_to_split=(0, 1)):
+def _ceil_power_of_two(x):
+    """The smallest power of two greater than or equal to `x`."""
+    return 2 ** int(np.ceil(np.log2(x)))
```

Comment:
This breaks down above about 2<sup>49</sup> due to rounding errors and for `x=0` since its log is -inf, but I assume neither should be an issue in practice.

Reply:
Indeed.